File: typescript/packages/dusk-adapter/src/codec.ts
import { createHash } from "node:crypto";

import type { InstructionContract } from "./envelope.ts";

export interface BidLiquidationAuctionArgs {
  readonly repayAmount: bigint;
  readonly minCollateralOut: bigint;
}

export interface SettleLiquidationAuctionFloorArgs {
  readonly repayAmount: bigint;
  readonly minCollateralOut: bigint;
  readonly maxInsuranceDraw: bigint;
  readonly maxSocializedLoss: bigint;
}

export interface LiquidateLeverageArgs {
  readonly debtAsset: number;
}

export interface DelegatedCpiArgs {
  readonly beforeIxData: Uint8Array;
  readonly afterIxData: Uint8Array;
  readonly beforeAccountsLen: number;
}

export interface DelegatedCloseLeverageArgs {
  readonly debtAsset: number;
  readonly minAmountOut: bigint;
  readonly delegated: DelegatedCpiArgs;
}

export interface ExecuteOrderArgs {
  readonly orderId: bigint;
}

export type ProtocolAuctionLane = "fee" | "buyback";
export type ProtocolRevenueSource = "swap" | "interest";

export interface SettleProtocolAuctionArgs {
  readonly lane: ProtocolAuctionLane;
  readonly source: ProtocolRevenueSource;
  readonly soldAmount: bigint;
  readonly maxPaymentAmount: bigint;
}

export type KeeperInstructionArguments =
  | { readonly specificationKey: "dusk:trigger_liquidation_auction" }
  | {
      readonly specificationKey: "dusk:bid_liquidation_auction";
      readonly arguments: BidLiquidationAuctionArgs;
    }
  | {
      readonly specificationKey: "dusk:settle_liquidation_auction_floor";
      readonly arguments: SettleLiquidationAuctionFloorArgs;
    }
  | {
      readonly specificationKey: "dusk:liquidate_leverage";
      readonly arguments: LiquidateLeverageArgs;
    }
  | {
      readonly specificationKey: "dusk:delegated_close_leverage";
      readonly arguments: DelegatedCloseLeverageArgs;
    }
  | {
      readonly specificationKey: "leverage_delegate:before_take_profit";
      readonly arguments: ExecuteOrderArgs;
    }
  | {
      readonly specificationKey: "leverage_delegate:after_close_order";
      readonly arguments: ExecuteOrderArgs;
    }
  | {
      readonly specificationKey: "leverage_delegate:before_stop_loss";
      readonly arguments: ExecuteOrderArgs;
    }
  | {
      readonly specificationKey: "dusk:settle_protocol_auction";
      readonly arguments: SettleProtocolAuctionArgs;
    }
  | { readonly specificationKey: "dusk:queue_parameter_proposal" }
  | { readonly specificationKey: "dusk:execute_parameter_proposal" };

export type InstructionEncodingErrorCode =
  | "unsupported_instruction"
  | "invalid_arguments"
  | "contract_mismatch";

export class InstructionEncodingError extends Error {
  readonly code: InstructionEncodingErrorCode;

  constructor(code: InstructionEncodingErrorCode, message: string) {
    super(message);
    this.name = "InstructionEncodingError";
    this.code = code;
  }
}

export function encodeKeeperInstruction(
  contract: InstructionContract,
  request: KeeperInstructionArguments,
): Uint8Array {
  const specification = contract.instructions.find(
    (entry) => entry.key === request.specificationKey,
  );
  if (!specification) {
    throw encodingError(
      "contract_mismatch",
      `${request.specificationKey}: instruction is absent from the pinned contract`,
    );
  }
  const [expectedProgram, expectedName, unexpectedPart] = request.specificationKey.split(":");
  if (
    unexpectedPart !== undefined ||
    specification.program !== expectedProgram ||
    specification.instructionName !== expectedName
  ) {
    throw encodingError(
      "contract_mismatch",
      `${request.specificationKey}: contract instruction identity differs`,
    );
  }
  const discriminator = exactHex(specification.discriminatorHex, 8, "contract_mismatch");
  const anchorDiscriminator = createHash("sha256")
    .update(`global:${specification.instructionName}`)
    .digest()
    .subarray(0, 8);
  if (!discriminator.equals(anchorDiscriminator)) {
    throw encodingError(
      "contract_mismatch",
      `${request.specificationKey}: discriminator differs from Anchor global hash`,
    );
  }

  const writer = new BorshWriter();
  switch (request.specificationKey) {
    case "dusk:trigger_liquidation_auction":
    case "dusk:queue_parameter_proposal":
    case "dusk:execute_parameter_proposal":
      break;
    case "dusk:bid_liquidation_auction":
      writer.u64(request.arguments.repayAmount);
      writer.u64(request.arguments.minCollateralOut);
      break;
    case "dusk:settle_liquidation_auction_floor":
      writer.u64(request.arguments.repayAmount);
      writer.u64(request.arguments.minCollateralOut);
      writer.u64(request.arguments.maxInsuranceDraw);
      writer.u64(request.arguments.maxSocializedLoss);
      break;
    case "dusk:liquidate_leverage":
      writer.u8(request.arguments.debtAsset);
      break;
    case "dusk:delegated_close_leverage":
      writer.u8(request.arguments.debtAsset);
      writer.u64(request.arguments.minAmountOut);
      writer.bytes(request.arguments.delegated.beforeIxData);
      writer.bytes(request.arguments.delegated.afterIxData);
      writer.u16(request.arguments.delegated.beforeAccountsLen);
      break;
    case "leverage_delegate:before_take_profit":
    case "leverage_delegate:after_close_order":
    case "leverage_delegate:before_stop_loss":
      writer.u64(request.arguments.orderId);
      break;
    case "dusk:settle_protocol_auction":
      writer.u8(enumOrdinal(request.arguments.lane, ["fee", "buyback"], "lane"));
      writer.u8(enumOrdinal(request.arguments.source, ["swap", "interest"], "source"));
      writer.u64(request.arguments.soldAmount);
      writer.u64(request.arguments.maxPaymentAmount);
      break;
  }
  return Buffer.concat([discriminator, writer.finish()]);
}

export function parseKeeperInstructionArguments(
  specificationKey: string,
  value: unknown,
): KeeperInstructionArguments {
  const argumentsValue = requireRecord(value);
  switch (specificationKey) {
    case "dusk:trigger_liquidation_auction":
    case "dusk:queue_parameter_proposal":
    case "dusk:execute_parameter_proposal":
      requireKeys(argumentsValue, []);
      return { specificationKey };
    case "dusk:bid_liquidation_auction":
      requireKeys(argumentsValue, ["repayAmount", "minCollateralOut"]);
      return {
        specificationKey,
        arguments: {
          repayAmount: decimalU64(argumentsValue.repayAmount),
          minCollateralOut: decimalU64(argumentsValue.minCollateralOut),
        },
      };
    case "dusk:settle_liquidation_auction_floor":
      requireKeys(argumentsValue, [
        "repayAmount",
        "minCollateralOut",
        "maxInsuranceDraw",
        "maxSocializedLoss",
      ]);
      return {
        specificationKey,
        arguments: {
          repayAmount: decimalU64(argumentsValue.repayAmount),
          minCollateralOut: decimalU64(argumentsValue.minCollateralOut),
          maxInsuranceDraw: decimalU64(argumentsValue.maxInsuranceDraw),
          maxSocializedLoss: decimalU64(argumentsValue.maxSocializedLoss),
        },
      };
    case "dusk:liquidate_leverage":
      requireKeys(argumentsValue, ["debtAsset"]);
      return {
        specificationKey,
        arguments: { debtAsset: unsignedNumber(argumentsValue.debtAsset, 0xff, "debtAsset") },
      };
    case "dusk:delegated_close_leverage":
      requireKeys(argumentsValue, [
        "debtAsset",
        "minAmountOut",
        "beforeIxDataHex",
        "afterIxDataHex",
        "beforeAccountsLen",
      ]);
      return {
        specificationKey,
        arguments: {
          debtAsset: unsignedNumber(argumentsValue.debtAsset, 0xff, "debtAsset"),
          minAmountOut: decimalU64(argumentsValue.minAmountOut),
          delegated: {
            beforeIxData: variableHex(argumentsValue.beforeIxDataHex),
            afterIxData: variableHex(argumentsValue.afterIxDataHex),
            beforeAccountsLen: unsignedNumber(
              argumentsValue.beforeAccountsLen,
              0xffff,
              "beforeAccountsLen",
            ),
          },
        },
      };
    case "leverage_delegate:before_take_profit":
    case "leverage_delegate:after_close_order":
    case "leverage_delegate:before_stop_loss":
      requireKeys(argumentsValue, ["orderId"]);
      return {
        specificationKey,
        arguments: { orderId: decimalU64(argumentsValue.orderId) },
      };
    case "dusk:settle_protocol_auction":
      requireKeys(argumentsValue, ["lane", "source", "soldAmount", "maxPaymentAmount"]);
      return {
        specificationKey,
        arguments: {
          lane: enumValue(argumentsValue.lane, ["fee", "buyback"], "lane"),
          source: enumValue(argumentsValue.source, ["swap", "interest"], "source"),
          soldAmount: decimalU64(argumentsValue.soldAmount),
          maxPaymentAmount: decimalU64(argumentsValue.maxPaymentAmount),
        },
      };
    default:
      throw encodingError(
        "unsupported_instruction",
        `${specificationKey}: unsupported keeper instruction`,
      );
  }
}

export function encodeKeeperInstructionJson(
  contract: InstructionContract,
  specificationKey: string,
  value: unknown,
): Uint8Array {
  return encodeKeeperInstruction(contract, parseKeeperInstructionArguments(specificationKey, value));
}

class BorshWriter {
  readonly #chunks: Buffer[] = [];

  u8(value: number): void {
    this.#unsigned(value, 0xff, 1, "u8");
  }

  u16(value: number): void {
    this.#unsigned(value, 0xffff, 2, "u16");
  }

  u32(value: number): void {
    this.#unsigned(value, 0xffff_ffff, 4, "u32");
  }

  u64(value: bigint): void {
    if (value < 0n || value > 0xffff_ffff_ffff_ffffn) {
      throw encodingError("invalid_arguments", "u64 value is outside its Borsh range");
    }
    const output = Buffer.alloc(8);
    output.writeBigUInt64LE(value);
    this.#chunks.push(output);
  }

  bytes(value: Uint8Array): void {
    if (value.byteLength > 0xffff_ffff) {
      throw encodingError("invalid_arguments", "Borsh byte vector exceeds u32 length");
    }
    this.u32(value.byteLength);
    this.#chunks.push(Buffer.from(value));
  }

  finish(): Buffer {
    return Buffer.concat(this.#chunks);
  }

  #unsigned(value: number, maximum: number, width: 1 | 2 | 4, label: string): void {
    if (!Number.isSafeInteger(value) || value < 0 || value > maximum) {
      throw encodingError("invalid_arguments", `${label} value is outside its Borsh range`);
    }
    const output = Buffer.alloc(width);
    if (width === 1) output.writeUInt8(value);
    if (width === 2) output.writeUInt16LE(value);
    if (width === 4) output.writeUInt32LE(value);
    this.#chunks.push(output);
  }
}

function requireRecord(value: unknown): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw encodingError("invalid_arguments", "instruction arguments must be an object");
  }
  return value as Record<string, unknown>;
}

function requireKeys(value: Record<string, unknown>, expected: readonly string[]): void {
  const actual = Object.keys(value).sort();
  const wanted = [...expected].sort();
  if (actual.length !== wanted.length || actual.some((entry, index) => entry !== wanted[index])) {
    throw encodingError("invalid_arguments", "instruction argument fields do not match the IDL");
  }
}

function decimalU64(value: unknown): bigint {
  if (
    typeof value !== "string" ||
    !/^(?:0|[1-9][0-9]*)$/.test(value)
  ) {
    throw encodingError("invalid_arguments", "u64 values must use canonical decimal strings");
  }
  const parsed = BigInt(value);
  if (parsed > 0xffff_ffff_ffff_ffffn) {
    throw encodingError("invalid_arguments", "u64 value is outside its Borsh range");
  }
  return parsed;
}

function unsignedNumber(value: unknown, maximum: number, name: string): number {
  if (
    typeof value !== "number" ||
    !Number.isSafeInteger(value) ||
    value < 0 ||
    value > maximum
  ) {
    throw encodingError("invalid_arguments", `${name} is outside its unsigned integer range`);
  }
  return value as number;
}

function enumOrdinal<const T extends string>(
  value: T,
  allowed: readonly T[],
  name: string,
): number {
  const ordinal = allowed.indexOf(value);
  if (ordinal < 0) {
    throw encodingError("invalid_arguments", `${name} is not a pinned IDL enum variant`);
  }
  return ordinal;
}

function enumValue<const T extends string>(
  value: unknown,
  allowed: readonly T[],
  name: string,
): T {
  if (typeof value !== "string" || !allowed.includes(value as T)) {
    throw encodingError("invalid_arguments", `${name} is not a pinned IDL enum variant`);
  }
  return value as T;
}

function variableHex(value: unknown): Buffer {
  if (typeof value !== "string") {
    throw encodingError("invalid_arguments", "byte vectors must use lowercase hex strings");
  }
  return exactHex(value, null, "invalid_arguments");
}

function exactHex(
  value: string,
  bytes: number | null,
  code: InstructionEncodingErrorCode,
): Buffer {
  if (!/^(?:[0-9a-f]{2})*$/.test(value) || (bytes !== null && value.length !== bytes * 2)) {
    throw encodingError(code, "value is not the expected even-length lowercase hex");
  }
  return Buffer.from(value, "hex");
}

function encodingError(
  code: InstructionEncodingErrorCode,
  message: string,
): InstructionEncodingError {
  return new InstructionEncodingError(code, message);
}

File: rust/crates/dusk-adapter/src/codec.rs
//! Native Borsh argument encoding for the keeper-critical instruction set.
//!
//! Decimal strings are used only at the JSON boundary so every u64 value is
//! lossless in both runtimes. Native callers use the strongly typed argument
//! structures below.

use std::{error::Error, fmt};

use borsh::BorshSerialize;
use serde::{Deserialize, Serialize, de::DeserializeOwned};
use serde_json::Value;
use sha2::{Digest, Sha256};

use crate::InstructionContract;

#[derive(Clone, Debug, BorshSerialize, Eq, PartialEq)]
pub struct BidLiquidationAuctionArgs {
    pub repay_amount: u64,
    pub min_collateral_out: u64,
}

#[derive(Clone, Debug, BorshSerialize, Eq, PartialEq)]
pub struct SettleLiquidationAuctionFloorArgs {
    pub repay_amount: u64,
    pub min_collateral_out: u64,
    pub max_insurance_draw: u64,
    pub max_socialized_loss: u64,
}

#[derive(Clone, Debug, BorshSerialize, Eq, PartialEq)]
pub struct LiquidateLeverageArgs {
    pub debt_asset: u8,
}

#[derive(Clone, Debug, BorshSerialize, Eq, PartialEq)]
pub struct DelegatedCpiArgs {
    pub before_ix_data: Vec<u8>,
    pub after_ix_data: Vec<u8>,
    pub before_accounts_len: u16,
}

#[derive(Clone, Debug, BorshSerialize, Eq, PartialEq)]
pub struct DelegatedCloseLeverageArgs {
    pub debt_asset: u8,
    pub min_amount_out: u64,
    pub delegated: DelegatedCpiArgs,
}

#[derive(Clone, Debug, BorshSerialize, Eq, PartialEq)]
pub struct ExecuteOrderArgs {
    pub order_id: u64,
}

#[derive(Clone, Copy, Debug, BorshSerialize, Eq, PartialEq)]
pub enum ProtocolAuctionLane {
    Fee,
    Buyback,
}

#[derive(Clone, Copy, Debug, BorshSerialize, Eq, PartialEq)]
pub enum ProtocolRevenueSource {
    Swap,
    Interest,
}

#[derive(Clone, Debug, BorshSerialize, Eq, PartialEq)]
pub struct SettleProtocolAuctionArgs {
    pub lane: ProtocolAuctionLane,
    pub source: ProtocolRevenueSource,
    pub sold_amount: u64,
    pub max_payment_amount: u64,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum KeeperInstructionArguments {
    TriggerLiquidationAuction,
    BidLiquidationAuction(BidLiquidationAuctionArgs),
    SettleLiquidationAuctionFloor(SettleLiquidationAuctionFloorArgs),
    LiquidateLeverage(LiquidateLeverageArgs),
    DelegatedCloseLeverage(DelegatedCloseLeverageArgs),
    BeforeTakeProfit(ExecuteOrderArgs),
    AfterCloseOrder(ExecuteOrderArgs),
    BeforeStopLoss(ExecuteOrderArgs),
    SettleProtocolAuction(SettleProtocolAuctionArgs),
    QueueParameterProposal,
    ExecuteParameterProposal,
}

impl KeeperInstructionArguments {
    pub fn specification_key(&self) -> &'static str {
        match self {
            Self::TriggerLiquidationAuction => "dusk:trigger_liquidation_auction",
            Self::BidLiquidationAuction(_) => "dusk:bid_liquidation_auction",
            Self::SettleLiquidationAuctionFloor(_) => "dusk:settle_liquidation_auction_floor",
            Self::LiquidateLeverage(_) => "dusk:liquidate_leverage",
            Self::DelegatedCloseLeverage(_) => "dusk:delegated_close_leverage",
            Self::BeforeTakeProfit(_) => "leverage_delegate:before_take_profit",
            Self::AfterCloseOrder(_) => "leverage_delegate:after_close_order",
            Self::BeforeStopLoss(_) => "leverage_delegate:before_stop_loss",
            Self::SettleProtocolAuction(_) => "dusk:settle_protocol_auction",
            Self::QueueParameterProposal => "dusk:queue_parameter_proposal",
            Self::ExecuteParameterProposal => "dusk:execute_parameter_proposal",
        }
    }

    fn borsh_bytes(&self) -> Result<Vec<u8>, InstructionEncodingError> {
        let result = match self {
            Self::TriggerLiquidationAuction
            | Self::QueueParameterProposal
            | Self::ExecuteParameterProposal => return Ok(Vec::new()),
            Self::BidLiquidationAuction(value) => borsh::to_vec(value),
            Self::SettleLiquidationAuctionFloor(value) => borsh::to_vec(value),
            Self::LiquidateLeverage(value) => borsh::to_vec(value),
            Self::DelegatedCloseLeverage(value) => borsh::to_vec(value),
            Self::BeforeTakeProfit(value)
            | Self::AfterCloseOrder(value)
            | Self::BeforeStopLoss(value) => borsh::to_vec(value),
            Self::SettleProtocolAuction(value) => borsh::to_vec(value),
        };
        result.map_err(|error| {
            InstructionEncodingError::new(
                InstructionEncodingErrorCode::InvalidArguments,
                format!("Borsh serialization failed: {error}"),
            )
        })
    }
}

pub fn encode_keeper_instruction(
    contract: &InstructionContract,
    arguments: &KeeperInstructionArguments,
) -> Result<Vec<u8>, InstructionEncodingError> {
    let specification_key = arguments.specification_key();
    let specification = contract
        .instructions
        .iter()
        .find(|entry| entry.key == specification_key)
        .ok_or_else(|| {
            InstructionEncodingError::new(
                InstructionEncodingErrorCode::ContractMismatch,
                format!("{specification_key}: instruction is absent from the pinned contract"),
            )
        })?;
    let (expected_program, expected_name) = specification_key.split_once(':').ok_or_else(|| {
        InstructionEncodingError::new(
            InstructionEncodingErrorCode::ContractMismatch,
            format!("{specification_key}: malformed specification key"),
        )
    })?;
    let actual_program = match specification.program {
        crate::ProgramName::Dusk => "dusk",
        crate::ProgramName::LeverageDelegate => "leverage_delegate",
    };
    if actual_program != expected_program || specification.instruction_name != expected_name {
        return Err(InstructionEncodingError::new(
            InstructionEncodingErrorCode::ContractMismatch,
            format!("{specification_key}: contract instruction identity differs"),
        ));
    }
    let discriminator = exact_hex::<8>(&specification.discriminator_hex).map_err(|message| {
        InstructionEncodingError::new(
            InstructionEncodingErrorCode::ContractMismatch,
            format!("{specification_key}: {message}"),
        )
    })?;
    let anchor_discriminator = Sha256::digest(format!("global:{}", specification.instruction_name));
    if discriminator != anchor_discriminator[..8] {
        return Err(InstructionEncodingError::new(
            InstructionEncodingErrorCode::ContractMismatch,
            format!("{specification_key}: discriminator differs from Anchor global hash"),
        ));
    }
    let mut output = discriminator.to_vec();
    output.extend(arguments.borsh_bytes()?);
    Ok(output)
}

/// Parses the lossless JSON wire form and returns native typed arguments.
pub fn parse_keeper_instruction_arguments(
    specification_key: &str,
    value: Value,
) -> Result<KeeperInstructionArguments, InstructionEncodingError> {
    match specification_key {
        "dusk:trigger_liquidation_auction" => {
            parse_empty(value)?;
            Ok(KeeperInstructionArguments::TriggerLiquidationAuction)
        }
        "dusk:bid_liquidation_auction" => {
            let wire: BidWire = parse_wire(value)?;
            Ok(KeeperInstructionArguments::BidLiquidationAuction(
                BidLiquidationAuctionArgs {
                    repay_amount: decimal_u64(&wire.repay_amount)?,
                    min_collateral_out: decimal_u64(&wire.min_collateral_out)?,
                },
            ))
        }
        "dusk:settle_liquidation_auction_floor" => {
            let wire: SettleFloorWire = parse_wire(value)?;
            Ok(KeeperInstructionArguments::SettleLiquidationAuctionFloor(
                SettleLiquidationAuctionFloorArgs {
                    repay_amount: decimal_u64(&wire.repay_amount)?,
                    min_collateral_out: decimal_u64(&wire.min_collateral_out)?,
                    max_insurance_draw: decimal_u64(&wire.max_insurance_draw)?,
                    max_socialized_loss: decimal_u64(&wire.max_socialized_loss)?,
                },
            ))
        }
        "dusk:liquidate_leverage" => {
            let wire: LiquidateWire = parse_wire(value)?;
            Ok(KeeperInstructionArguments::LiquidateLeverage(
                LiquidateLeverageArgs {
                    debt_asset: wire.debt_asset,
                },
            ))
        }
        "dusk:delegated_close_leverage" => {
            let wire: DelegatedCloseWire = parse_wire(value)?;
            Ok(KeeperInstructionArguments::DelegatedCloseLeverage(
                DelegatedCloseLeverageArgs {
                    debt_asset: wire.debt_asset,
                    min_amount_out: decimal_u64(&wire.min_amount_out)?,
                    delegated: DelegatedCpiArgs {
                        before_ix_data: variable_hex(&wire.before_ix_data_hex)?,
                        after_ix_data: variable_hex(&wire.after_ix_data_hex)?,
                        before_accounts_len: wire.before_accounts_len,
                    },
                },
            ))
        }
        "leverage_delegate:before_take_profit" => {
            parse_order(value).map(KeeperInstructionArguments::BeforeTakeProfit)
        }
        "leverage_delegate:after_close_order" => {
            parse_order(value).map(KeeperInstructionArguments::AfterCloseOrder)
        }
        "leverage_delegate:before_stop_loss" => {
            parse_order(value).map(KeeperInstructionArguments::BeforeStopLoss)
        }
        "dusk:settle_protocol_auction" => {
            let wire: SettleProtocolWire = parse_wire(value)?;
            Ok(KeeperInstructionArguments::SettleProtocolAuction(
                SettleProtocolAuctionArgs {
                    lane: wire.lane.into(),
                    source: wire.source.into(),
                    sold_amount: decimal_u64(&wire.sold_amount)?,
                    max_payment_amount: decimal_u64(&wire.max_payment_amount)?,
                },
            ))
        }
        "dusk:queue_parameter_proposal" => {
            parse_empty(value)?;
            Ok(KeeperInstructionArguments::QueueParameterProposal)
        }
        "dusk:execute_parameter_proposal" => {
            parse_empty(value)?;
            Ok(KeeperInstructionArguments::ExecuteParameterProposal)
        }
        _ => Err(InstructionEncodingError::new(
            InstructionEncodingErrorCode::UnsupportedInstruction,
            format!("{specification_key}: unsupported keeper instruction"),
        )),
    }
}

pub fn encode_keeper_instruction_json(
    contract: &InstructionContract,
    specification_key: &str,
    value: Value,
) -> Result<Vec<u8>, InstructionEncodingError> {
    let arguments = parse_keeper_instruction_arguments(specification_key, value)?;
    encode_keeper_instruction(contract, &arguments)
}

#[derive(Clone, Copy, Debug, Deserialize)]
#[serde(rename_all = "snake_case")]
enum ProtocolAuctionLaneWire {
    Fee,
    Buyback,
}

impl From<ProtocolAuctionLaneWire> for ProtocolAuctionLane {
    fn from(value: ProtocolAuctionLaneWire) -> Self {
        match value {
            ProtocolAuctionLaneWire::Fee => Self::Fee,
            ProtocolAuctionLaneWire::Buyback => Self::Buyback,
        }
    }
}

#[derive(Clone, Copy, Debug, Deserialize)]
#[serde(rename_all = "snake_case")]
enum ProtocolRevenueSourceWire {
    Swap,
    Interest,
}

impl From<ProtocolRevenueSourceWire> for ProtocolRevenueSource {
    fn from(value: ProtocolRevenueSourceWire) -> Self {
        match value {
            ProtocolRevenueSourceWire::Swap => Self::Swap,
            ProtocolRevenueSourceWire::Interest => Self::Interest,
        }
    }
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct BidWire {
    repay_amount: String,
    min_collateral_out: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct SettleFloorWire {
    repay_amount: String,
    min_collateral_out: String,
    max_insurance_draw: String,
    max_socialized_loss: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct LiquidateWire {
    debt_asset: u8,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct DelegatedCloseWire {
    debt_asset: u8,
    min_amount_out: String,
    before_ix_data_hex: String,
    after_ix_data_hex: String,
    before_accounts_len: u16,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct OrderWire {
    order_id: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct SettleProtocolWire {
    lane: ProtocolAuctionLaneWire,
    source: ProtocolRevenueSourceWire,
    sold_amount: String,
    max_payment_amount: String,
}

fn parse_empty(value: Value) -> Result<(), InstructionEncodingError> {
    #[derive(Deserialize)]
    #[serde(deny_unknown_fields)]
    struct EmptyArguments {}
    parse_wire::<EmptyArguments>(value).map(|_| ())
}

fn parse_order(value: Value) -> Result<ExecuteOrderArgs, InstructionEncodingError> {
    let wire: OrderWire = parse_wire(value)?;
    Ok(ExecuteOrderArgs {
        order_id: decimal_u64(&wire.order_id)?,
    })
}

fn parse_wire<T: DeserializeOwned>(value: Value) -> Result<T, InstructionEncodingError> {
    serde_json::from_value(value).map_err(|error| {
        InstructionEncodingError::new(
            InstructionEncodingErrorCode::InvalidArguments,
            format!("invalid instruction arguments: {error}"),
        )
    })
}

fn decimal_u64(value: &str) -> Result<u64, InstructionEncodingError> {
    if value.is_empty()
        || (value != "0" && value.starts_with('0'))
        || !value.bytes().all(|byte| byte.is_ascii_digit())
    {
        return Err(invalid_argument(
            "u64 values must use canonical decimal strings",
        ));
    }
    value
        .parse::<u64>()
        .map_err(|_| invalid_argument("u64 value is outside 0..=18446744073709551615"))
}

fn variable_hex(value: &str) -> Result<Vec<u8>, InstructionEncodingError> {
    if !value.len().is_multiple_of(2) {
        return Err(invalid_argument(
            "byte vectors must use even-length lowercase hex",
        ));
    }
    decode_hex_bytes(value).map_err(invalid_argument)
}

fn exact_hex<const N: usize>(value: &str) -> Result<[u8; N], &'static str> {
    let bytes = decode_hex_bytes(value)?;
    bytes
        .try_into()
        .map_err(|_| "discriminator does not contain exactly eight bytes")
}

fn decode_hex_bytes(value: &str) -> Result<Vec<u8>, &'static str> {
    if !value.len().is_multiple_of(2)
        || !value
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
    {
        return Err("value is not even-length lowercase hex");
    }
    value
        .as_bytes()
        .chunks_exact(2)
        .map(|pair| {
            let high = hex_nibble(pair[0]).ok_or("value is not lowercase hex")?;
            let low = hex_nibble(pair[1]).ok_or("value is not lowercase hex")?;
            Ok((high << 4) | low)
        })
        .collect()
}

fn hex_nibble(value: u8) -> Option<u8> {
    match value {
        b'0'..=b'9' => Some(value - b'0'),
        b'a'..=b'f' => Some(value - b'a' + 10),
        _ => None,
    }
}

fn invalid_argument(message: impl Into<String>) -> InstructionEncodingError {
    InstructionEncodingError::new(InstructionEncodingErrorCode::InvalidArguments, message)
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum InstructionEncodingErrorCode {
    UnsupportedInstruction,
    InvalidArguments,
    ContractMismatch,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct InstructionEncodingError {
    pub code: InstructionEncodingErrorCode,
    message: String,
}

impl InstructionEncodingError {
    fn new(code: InstructionEncodingErrorCode, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
        }
    }
}

impl fmt::Display for InstructionEncodingError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl Error for InstructionEncodingError {}

#[cfg(test)]
mod tests {
    use super::*;

    const CONTRACT: &str = include_str!("../../../../protocol/keeper-instructions.v1.json");
    const FIXTURES: &str =
        include_str!("../../../../fixtures/conformance/v1/adapter-codec-cases.json");

    #[derive(Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct FixtureBundle {
        encoding_cases: Vec<EncodingCase>,
        invalid_encoding_cases: Vec<InvalidEncodingCase>,
    }

    #[derive(Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct EncodingCase {
        name: String,
        specification_key: String,
        arguments: Value,
        expected_data_hex: String,
    }

    #[derive(Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct InvalidEncodingCase {
        name: String,
        specification_key: String,
        arguments: Value,
        expected_error: InstructionEncodingErrorCode,
    }

    fn contract() -> InstructionContract {
        InstructionContract::from_json(CONTRACT).expect("instruction contract must parse")
    }

    #[test]
    fn native_borsh_encoding_matches_every_shared_golden_vector() {
        let fixtures: FixtureBundle =
            serde_json::from_str(FIXTURES).expect("adapter codec fixtures must parse");
        assert_eq!(fixtures.encoding_cases.len(), 11);
        for fixture in fixtures.encoding_cases {
            let encoded = encode_keeper_instruction_json(
                &contract(),
                &fixture.specification_key,
                fixture.arguments,
            )
            .unwrap_or_else(|error| panic!("{}: {error}", fixture.name));
            assert_eq!(hex(&encoded), fixture.expected_data_hex, "{}", fixture.name);
        }
    }

    #[test]
    fn rejects_every_shared_invalid_argument_vector() {
        let fixtures: FixtureBundle =
            serde_json::from_str(FIXTURES).expect("adapter codec fixtures must parse");
        for fixture in fixtures.invalid_encoding_cases {
            let error = encode_keeper_instruction_json(
                &contract(),
                &fixture.specification_key,
                fixture.arguments,
            )
            .unwrap_err();
            assert_eq!(error.code, fixture.expected_error, "{}", fixture.name);
        }
    }

    #[test]
    fn rejects_a_discriminator_drift_in_the_contract() {
        let mut contract = contract();
        contract.instructions[0].discriminator_hex = "00".repeat(8);
        let error = encode_keeper_instruction(
            &contract,
            &KeeperInstructionArguments::TriggerLiquidationAuction,
        )
        .expect_err("a contract mismatch must fail closed");
        assert_eq!(error.code, InstructionEncodingErrorCode::ContractMismatch);
    }

    fn hex(bytes: &[u8]) -> String {
        const DIGITS: &[u8; 16] = b"0123456789abcdef";
        let mut output = String::with_capacity(bytes.len() * 2);
        for byte in bytes {
            output.push(DIGITS[(byte >> 4) as usize] as char);
            output.push(DIGITS[(byte & 0x0f) as usize] as char);
        }
        output
    }
}
